Updated H⁻ bf opacity calculation

mabruzzo opened the pull request:
I altered `ContinuumOpacity.Hminus_bf` so that the cross-section is computed by directly interpolating data from Wishart (1979).

Previously, the cross-sections were computed using a polynomial from Gray (2005) that was fit to the tabulated data in Wishart (1979) while the tabulated data was used for verify the polynomial. Now, we use the Gray (2005) polynomial in our test to confirm that our transcription of the table is correct.

This change was primarily motivated by the fact directly using the table gives us opacities to wavelengths that are 1300 Å than before. 

While looking at [Mathisen (1984) ](https://inis.iaea.org/search/search.aspx?orig_q=RN:16033032), it clicked that the cross-section obviously goes to 0 at the ionization energy of H⁻.
- In other words,  there is no longer an upper wavelength limit where we don't know the H⁻ bound free opacity
- for reference, this zero-point is ~140 Å above the max wavelength in the table (in a part of the table where the wavelength spacing is 100 Å)
- my implementation for handling this is definitely not efficient. But I'm inclined to keep it until there is evidence that it is a performance bottleneck

I'm pretty confident that this addresses the IR solar spectrum discrepancy with `MOOG` that was described in PR #50. Here is the updated continuum spectrum:

![solar_comparison](https://user-images.githubusercontent.com/28722054/123729954-c39adc00-d863-11eb-8049-2ba3b97c6f6b.png)

I wanted to note one thing about what `MOOG`'s handling of the H⁻ bf opacity:
- They follow the suggestion of [Mathisen (1984) ](https://inis.iaea.org/search/search.aspx?orig_q=RN:16033032), and use a different data source for wavelengths from 180 Å to 1640 Å. 
    - My inclination is not to do this, since it doesn't really get us a whole lot since our current source gets down to 1250 Å. I could add a note about this to Issue #12 in case there's ever any interest in sythesizing UV spectra.
    - Moreover, there are apparently some resonance effects below 1250 Å, and that makes the accuracy of the other source a little questionable.

ajwheeler commented on src/continuum_opacity/opacity_H.jl:
```diff
@@ -44,29 +44,56 @@ function _ndens_Hminus(nH_I_div_partition, ne, T, ion_energy = _H⁻_ion_energy)
+    Å_per_eV = 1e8 * (hplanck_eV * c_cgs)
+    max_λ_ionize = Å_per_eV/ion_energy_H⁻
+
+    last_table_λ = _Hminus_bf_table[size(_Hminus_bf_table, 1), 1]
+    last_table_cross_section = _Hminus_bf_table[size(_Hminus_bf_table, 1), 2]
```
```suggestion
    last_table_λ = _Hminus_bf_table[end, 1]
    last_table_cross_section = _Hminus_bf_table[end, 2]
```

ajwheeler commented on src/continuum_opacity/opacity_H.jl:
```diff
@@ -44,29 +44,56 @@ function _ndens_Hminus(nH_I_div_partition, ne, T, ion_energy = _H⁻_ion_energy)
+
+    @assert max_λ_ionize > last_table_λ
+
+    if λ < _Hminus_bf_table[1,1]
+        throw(DomainError(λ, "$(_Hminus_bf_table[1,1]) Å must not exceed the wavelength"))
```
Is this `>` the right way around?

ajwheeler replied:
Or the word "exceed" wrong?

mabruzzo replied:
This was technically correct as written, but it definitely looked weird (I basically flipped the inequality). I changed it so that it makes more sense.

ajwheeler replied:
Ooooooh, this took me way longer than it should have.

ajwheeler commented on src/continuum_opacity/opacity_H.jl:
```diff
@@ -88,11 +115,14 @@ Compute the H⁻ bound-free opacity κ
 pre-computed (instead it's computed internally by this function).
 
-This function is adapted from equation 8.12 from Grey (2005). This equation gives the absorption
+This function is adapted from equation 8.12 from Grey (2005) This equation gives the absorption
```
```suggestion
This function is adapted from equation 8.12 from Grey (2005). This equation gives the absorption
```

ajwheeler commented on src/continuum_opacity/opacity_H.jl:
```diff
@@ -101,26 +131,23 @@ Eqn 8.12 of Grey (2005) implicitly assumes that (U(H⁻,T)/ U(H I,T)) is always
-    αbf_H⁻ * n_H⁻ * stimulated_emission_correction / ρ
+    λ = c_cgs*1e8/ν # in ångstroms
+    tmp = _Hminus_bf_cross_section(λ, ion_energy_H⁻) # in units of megabarn
+    # convert from megabarn to cm² and include contributions from stimulated emission  1e-18
+    αbf_H⁻ = (1 - exp(-hplanck_cgs*ν/(kboltz_cgs*T))) * tmp * 1e-18
```
Do you have a source for the `1e18`?

mabruzzo replied:
There are 1e18 megabarn in 1 cm². I'll clean up the code to be more clear

ajwheeler replied:
This is fine as is.  It was a thinko on my part.

ajwheeler commented on src/continuum_opacity/opacity_H.jl:
```diff
@@ -101,26 +131,23 @@ Eqn 8.12 of Grey (2005) implicitly assumes that (U(H⁻,T)/ U(H I,T)) is always
-    n_H⁻ = _ndens_Hminus(nH_I_div_partition, ne, T, _H⁻_ion_energy)
-    αbf_H⁻ * n_H⁻ * stimulated_emission_correction / ρ
+    λ = c_cgs*1e8/ν # in ångstroms
+    tmp = _Hminus_bf_cross_section(λ, ion_energy_H⁻) # in units of megabarn
+    # convert from megabarn to cm² and include contributions from stimulated emission  1e-18
+    αbf_H⁻ = (1 - exp(-hplanck_cgs*ν/(kboltz_cgs*T))) * tmp * 1e-18
+    _ndens_Hminus(nH_I_div_partition, ne, T, _H⁻_ion_energy) * αbf_H⁻ / ρ
```
`αbf_H⁻` isn't really an alpha value is it?

```suggestion
    cross_section = _Hminus_bf_cross_section(λ, ion_energy_H⁻) # in units of megabarn
    # convert from megabarn to cm² and include contributions from stimulated emission  1e-18
    cross_section *= (1 - exp(-hplanck_cgs*ν/(kboltz_cgs*T))) * 1e-18
    _ndens_Hminus(nH_I_div_partition, ne, T, _H⁻_ion_energy) * cross_section / ρ
```

mabruzzo replied:
This is the Gray (2005) notation. They use α to refer to the cross-section (unlike Rybicki & Lightman who I think use σ?). I'll definitely make this change, but it's something to keep in mind documentation-wise.

ajwheeler replied:
Ah yes, sorry I keep forgetting that.

ajwheeler commented:
Ah, that was a thinko on my part. I don’t think you need to
change anything.

On Tue, Jun 29, 2021 at 8:49 AM Matthew Abruzzo ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/continuum_opacity/opacity_H.jl
> <https://github.com/ajwheeler/Korg.jl/pull/51#discussion_r660583398>:
>
> > +    # convert from megabarn to cm² and include contributions from stimulated emission  1e-18
>
> +    αbf_H⁻ = (1 - exp(-hplanck_cgs*ν/(kboltz_cgs*T))) * tmp * 1e-18
>
>
> There are 1e18 megabarn in 1 cm². I'll clean up the code to be more clear
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/ajwheeler/Korg.jl/pull/51#discussion_r660583398>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFN2G4XKVDOOBXETTKQOOTTVG6ORANCNFSM47PFOZDA>
> .
>

mabruzzo commented:
Minor heads up: I think I introduced a new test set to improve test coverage. I think I pushed that just after you finished your review, so you may have missed it. I think I addressed all of your other concerns.

ajwheeler commented:
![image](https://user-images.githubusercontent.com/711963/123802766-a42e9e00-d8b9-11eb-8877-4c4792ed1dd5.png)

ajwheeler commented on src/continuum_opacity/opacity_H.jl:
```diff
@@ -44,29 +44,56 @@ function _ndens_Hminus(nH_I_div_partition, ne, T, ion_energy = _H⁻_ion_energy)
+    @assert max_λ_ionize > last_table_λ
+
+    if λ < _Hminus_bf_table[1,1]
+        throw(DomainError(λ, "$(_Hminus_bf_table[1,1]) Å must not exceed the wavelength"))
```
```suggestion
        throw(DomainError(λ, "λ must be ≥ $(_Hminus_bf_table[1,1]) Å"))
```

mabruzzo replied:
I'll apply this later (I can't apply it easily now since you didn't make the suggestion using the latest changeset)

ajwheeler commented on src/continuum_opacity/opacity_H.jl:
```diff
@@ -44,29 +44,56 @@ function _ndens_Hminus(nH_I_div_partition, ne, T, ion_energy = _H⁻_ion_energy)
+    @assert max_λ_ionize > last_table_λ
+
+    if λ < _Hminus_bf_table[1,1]
+        throw(DomainError(λ, "wavelength must be greater or equal to $(_Hminus_bf_table[1,1]) Å."))
```
```suggestion
        throw(DomainError(λ, "λ must be ≥ $(_Hminus_bf_table[1,1]) Å"))
```

ajwheeler replied:
I think I did it right this time.
